module/zfs/vdev_raidz.c

```diff
@@ -2036,22 +2082,30 @@ vdev_raidz_io_start(zio_t *zio)
 	raidz_map_t *rm;
 
 	if (vdrz->vd_logical_width != vdrz->vd_physical_width) {
-		/* XXX rangelock not needed after expansion completes */
-		zfs_locked_range_t *lr =
-		    zfs_rangelock_enter(&vdrz->vn_vre.vre_rangelock,
-		    zio->io_offset, zio->io_size, RL_READER);
-
-		rm = vdev_raidz_map_alloc_expanded(zio->io_abd,
-		    zio->io_size, zio->io_offset,
-		    tvd->vdev_ashift, vdrz->vd_physical_width,
-		    vdrz->vd_logical_width, vdrz->vd_nparity,
-		    vdrz->vn_vre.vre_offset_phys);
-		rm->rm_lr = lr;
-		/*
-		 * XXX If this is a write, will need to do additional
-		 * writes to locations that are already copied, but
-		 * not yet reflected in the on-disk format.
-		 */
+		uint64_t width = vdev_raidz_get_width(vdrz, zio->io_bp->blk_birth);
+		if (vdrz->vn_vre.vre_offset != UINT64_MAX ||
+		    (zio->io_type == ZIO_TYPE_READ && width != vdrz->vd_physical_width)) {
```

Comment:
If we could arrange for `vdev_raidz_get_width()` to tell us what width a block with the given birth time is stored in, then I think we should be able to simplify this to:

`if (width != vdrz->vd_physical_width) {`

Since I think that writes would always be using the "current" txg.

That said, we'd want to think carefully about how this interacts with ZIL writes, which might be happening with a "future" txg?

Reply:
If am not wrong it is needed leave vdrz->vn_vre.vre_offset != UINT64_MAX check, because we do not want to write BP's in regular format during reflow.